nixos/modules/services/misc/mautrix-telegram.nix

```diff
@@ -0,0 +1,145 @@
+        '';
+      };
+
+      environmentFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          File containing environment variables to be passed to the mautrix-telegram service,
+          in which secret tokens can be specified securely by defining values for
+          <literal>MAUTRIX_TELEGRAM_APPSERVICE_AS_TOKEN</literal>,
+          <literal>MAUTRIX_TELEGRAM_APPSERVICE_HS_TOKEN</literal>,
+          <literal>MAUTRIX_TELEGRAM_TELEGRAM_API_ID</literal>,
+          <literal>MAUTRIX_TELEGRAM_TELEGRAM_API_HASH</literal> and optionally
+          <literal>MAUTRIX_TELEGRAM_TELEGRAM_BOT_TOKEN</literal>.
+        '';
+      };
+
+      serviceDependencies = mkOption {
+        type = with types; listOf str;
+        default = [ ];
+        example = literalExample ''
+          [ "matrix-synapse.service" ]
+        '';
+        description = ''
+          List of Systemd services to require and await for when starting the application service.
+        '';
+      };
+    };
+  };
+
+  config = let
+    dataDir = "/var/lib/mautrix-telegram";
+    cfg = config.services.mautrix-telegram;
+  in mkIf cfg.enable {
+
+    # default minimal configuration
+    services.mautrix-telegram.config = {
+      appservice = rec {
+        database = mkDefault "sqlite:///${dataDir}/mautrix-telegram.db";
+        hostname = mkDefault "0.0.0.0";
+        port = mkDefault 8080;
+        address = mkDefault "http://localhost:${toString port}";
+      };
+
+      bridge = {
+        permissions."*" = mkDefault "relaybot";
+        relaybot.whitelist = mkDefault [ ];
+      };
+
+      # log to console/systemd instead of file
+      logging = {
+        version = 1;
+
+        formatters.precise.format = "[%(levelname)s@%(name)s] %(message)s";
+
+        handlers.console = {
+          class = "logging.StreamHandler";
+          formatter = "precise";
+        };
+
+        loggers = {
+          mau.level = mkDefault "INFO";
+          telethon.level = mkDefault "INFO";
+          aiohttp.level = mkDefault "INFO";
+        };
+
+        root = {
+          level = mkDefault "INFO";
+          handlers = [ "console" ];
+        };
+      };
+    };
+
+    systemd.services.mautrix-telegram = let
+      serviceConfig = pkgs.writeText "mautrix-telegram-config.json" (builtins.toJSON cfg.config);
+    in {
+      description = "Mautrix-Telegram, a Matrix-Telegram hybrid puppeting/relaybot bridge.";
+
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ] ++ cfg.serviceDependencies;
+      after = [ "network-online.target" ] ++ cfg.serviceDependencies;
+
+      preStart = ''
+        # run automatic database init and migration scripts
+        ${pkgs.mautrix-telegram.alembic}/bin/alembic -x config='${serviceConfig}' upgrade head
+      '';
+
+      serviceConfig = {
+        DynamicUser = true;
+        StateDirectory = baseNameOf dataDir;
```

Comment:
I think this should be an option in the module configuration aswell. I wouldn't expect something creating a file outside its state directory.

Reply:
`StateDirectory = baseNameOf dataDir` makes systemd create `/var/lib/mautrix-telegram`.
Everything is contained within that directory.